### PR TITLE
Update end time of last segment in a period when new periods are added

### PR DESCRIPTION
**Pull request by cmgrecu**
Hi,
I have a multi period live stream and, at some time shaka was stopping with an INVALID_SEGMENT_INDEX error.

I discovered that the issue was caused by the fact that when segment references are merged at the manifest change, some of the segments are skipped, considered to be duplicates of the previous received segments. This is true in general, but when the period changes, last segment is expanded to end of the period (in MpdUtils.fitSegmentReferences). This updated segment was not merged into the final segment list which was causing a gap.

I added a unit test that reproduce the issue and I also made a fix for it.

### CODE REVIEW

File: lib/media/segment_index.js [hunk 1]
```diff
@@ -129,9 +129,24 @@ shaka.media.SegmentIndex.prototype.merge = function(references) {
...
+      // expand last segment to the start of the next the period
+      // so, it is valid to have end time updated to the
+      // last segment reference in a period
+      if (r1.endTime != r2.endTime)
+      {
+        goog.asserts.assert(r2.endTime - r1.endTime <=
+            shaka.dash.MpdUtils.GAP_OVERLAP_TOLERANCE_SECONDS &&
+            i == this.references_.length - 1 &&
+            j == references.length - 1,
+            'This should be un update of the last segment in a period');
```
Comment (joeyparrish) [inline lib/media/segment_index.js]: typo: un => an

File: lib/media/segment_index.js [hunk 2]
```diff
@@ -129,9 +129,24 @@ shaka.media.SegmentIndex.prototype.merge = function(references) {
...
     } else {
-      // Drop the new reference if there's an old reference with the same
-      // time.
-      newReferences.push(r1);
+      // when period is changed, fitSegmentReference will
+      // expand last segment to the start of the next the period
+      // so, it is valid to have end time updated to the
+      // last segment reference in a period
+      if (r1.endTime != r2.endTime)
+      {
```
Comment (joeyparrish) [inline lib/media/segment_index.js]: Our style convention is to put curly braces on the previous line.  Please update.

File: lib/media/segment_index.js [hunk 3]
```diff
@@ -129,9 +129,24 @@ shaka.media.SegmentIndex.prototype.merge = function(references) {
...
+        goog.asserts.assert(r2.endTime - r1.endTime <=
+            shaka.dash.MpdUtils.GAP_OVERLAP_TOLERANCE_SECONDS &&
+            i == this.references_.length - 1 &&
+            j == references.length - 1,
+            'This should be un update of the last segment in a period');
+        newReferences.push(r2);
+      }
+      else {
```
Comment (joeyparrish) [inline lib/media/segment_index.js]: Please move else to the previous line to match the style in the rest of the project:

} else {

File: lib/media/segment_index.js [hunk 4]
```diff
@@ -129,9 +129,24 @@ shaka.media.SegmentIndex.prototype.merge = function(references) {
       shaka.log.warning('Refusing to rewrite original references on update!');
       j++;
     } else {
-      // Drop the new reference if there's an old reference with the same
-      // time.
-      newReferences.push(r1);
+      // when period is changed, fitSegmentReference will
+      // expand last segment to the start of the next the period
```
Comment (tdrews) [inline lib/media/segment_index.js]: typo: double "the".

File: lib/media/segment_index.js [hunk 5]
```diff
@@ -129,9 +129,24 @@ shaka.media.SegmentIndex.prototype.merge = function(references) {
...
+      // so, it is valid to have end time updated to the
+      // last segment reference in a period
+      if (r1.endTime != r2.endTime)
+      {
+        goog.asserts.assert(r2.endTime - r1.endTime <=
```
Comment (tdrews) [inline lib/media/segment_index.js]: Suggest that you also assert that r2.endTime > r1.endTime.

**Comment by cmgrecu** [comment]
@joeyparrish @tdrews 
Thanks for feedback.
Done! I hope I didn't miss anything.

**Comment by cmgrecu** [comment]
Done.
Thank you.
Sorry for so much typos in 5 lines of code :)

**Comment by joeyparrish** [comment]
No worries!  I'll run it through our buildbot now for one last check before we merge.  Thanks for contributing!

**Pull request merged**